pkg/util/rbac/rbac.go

```diff
@@ -68,32 +68,90 @@ func ResourceRoleAssignmentWithName(roleID, spID, resourceType, resourceName, na
+	var resourceIdFunction string
+	if subscriptionScope {
+		resourceIdFunction = "subscriptionResourceId"
+	} else {
+		resourceIdFunction = "resourceId"
+	}
 	r := &arm.Resource{
 		Resource: mgmtauthorization.RoleAssignment{
 			Name: to.StringPtr("[" + name + "]"),
 			Type: to.StringPtr("Microsoft.Authorization/roleAssignments"),
 			RoleAssignmentPropertiesWithScope: &mgmtauthorization.RoleAssignmentPropertiesWithScope{
 				Scope:            to.StringPtr("[resourceGroup().id]"),
-				RoleDefinitionID: to.StringPtr("[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]"),
+				RoleDefinitionID: to.StringPtr("[" + resourceIdFunction + "('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]"),
 				PrincipalID:      to.StringPtr("[" + spID + "]"),
 				PrincipalType:    mgmtauthorization.ServicePrincipal,
 			},
 		},
 		APIVersion: azureclient.APIVersion("Microsoft.Authorization"),
 	}
-
+	if dependsOn != nil {
+		r.DependsOn = dependsOn
+	}
 	if len(condition) > 0 {
 		r.Condition = condition[0]
 	}
+	return r
+}
+
+// ResourceGroupRoleAssignmentWithName returns a Resource granting roleID on the
+// current resource group to spID.  Arguments spID and name must be valid ARM
+// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
+// templates which have already been deployed, to preserve the name and avoid a
+// RoleAssignmentExists error.
+func ResourceGroupRoleAssignmentWithName(roleID, spID string, name string, condition ...interface{}) *arm.Resource {
+	return resourceGroupRoleAssignmentWithDetails(roleID, spID, name, nil, true, condition...)
+}
+
+// ResourceGroupCustomRoleAssignmentWithName returns a Resource granting roleID on the
+// current resource group to spID.  Arguments spID and name must be valid ARM
+// expressions, e.g. "'foo'" or "concat('foo')".  Use this function in ARM
+// templates which have already been deployed, to preserve the name and avoid a
+// RoleAssignmentExists error. Additionally a dependency is added on the role, use this function
+// when the Role Definition is declared in the same ARM template at non-subscription scope.
+func ResourceGroupCustomRoleAssignmentWithName(roleID, spID string, name string, condition ...interface{}) *arm.Resource {
+	var dependsOn []string
+	dependsOn = []string{
+		"[resourceId('Microsoft.Authorization/roleDefinitions', '" + roleID + "')]",
+	}
+	return resourceGroupRoleAssignmentWithDetails(roleID, spID, name, dependsOn, false, condition...)
+}
+
+// ResourceRoleDefinitionWithName returns a Resource defining a role granting the required
+// actions and dataActions, forbidding notActions and notDataActions. Argument name must be valid ARM expressions.
+func ResourceRoleDefinitionWithName(name string, actions, notActions, dataActions, notDataActions *[]string) *arm.Resource {
```

Comment:
ideally I would rather this also not end in WithName.  I recognise you need to get the name for the role assignment.  Perhaps an additional exported function which returns the string "guid(...)" and is called in both places?

Reply:
Changed to "ResourceRoleDefinition" and also "ResourceGroupCustomRoleAssignmentWithName" to "ResourceGroupCustomRoleAssignment". WithName doesn't differentiate from other functions in this context.